server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -73,21 +78,31 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<ReplicaShardRefreshRequest, ReplicationResponse>> listener
     ) {
         ActionListener.completeWith(listener, () -> {
-            primary.refresh(SOURCE_API);
+            var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            return new PrimaryResult<>(shardRequest, new ReplicationResponse());
+            var shardRefreshRequest = new ReplicaShardRefreshRequest(
+                primary.shardId(),
+                shardRequest.getParentTask(),
+                refreshResult.generation()
+            );
+            return new PrimaryResult<>(shardRefreshRequest, new ReplicationResponse());
         });
     }
 
     @Override
-    protected void shardOperationOnReplica(BasicReplicationRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        ActionListener.completeWith(listener, () -> {
-            replica.refresh(SOURCE_API);
-            logger.trace("{} refresh request executed on replica", replica.shardId());
-            return new ReplicaResult();
-        });
+    protected void shardOperationOnReplica(ReplicaShardRefreshRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
+        if (DiscoveryNode.isStateless(settings) && replica.routingEntry().isPromotableToPrimary() == false) {
```

Comment:
Hi! I'm reviewing changes and wondering if you'd have time to explain what is the purpose of the `isPromotableToPrimary` in the condition? Would help me a lot understand the details. Thanks

Reply:
It should help distinguish between index and search shards in a more abstract way. I think this would be helpful later on where there might be more than one index shard (i.e. we'd be in `shardOperationOnReplica` but we are not a search shard). The default role (in stateful) is always promotable to primary which means we'd do the "normal refresh" on the shard. Only a search shard is never promotable to primary (i.e. does indexing) and refresh in this case would just mean waiting to see the changes.